src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs

```diff
@@ -0,0 +1,630 @@
+        {
+            TarHeader geaHeader = default;
+            geaHeader._name = GenerateGlobalExtendedAttributeName();
+            geaHeader._mode = (int)TarHelpers.DefaultMode;
+            geaHeader._typeFlag = TarEntryType.GlobalExtendedAttributes;
+            geaHeader._linkName = string.Empty;
+            geaHeader._magic = string.Empty;
+            geaHeader._version = string.Empty;
+            geaHeader._gName = string.Empty;
+            geaHeader._uName = string.Empty;
+            geaHeader.WriteAsPaxExtendedAttributes(archiveStream, buffer, globalExtendedAttributes, isGea: true);
+        }
+
+        // Writes the current header as a V7 entry into the archive stream.
+        internal void WriteAsV7(Stream archiveStream, Span<byte> buffer)
+        {
+            long actualLength = GetTotalDataBytesToWrite();
+            TarEntryType actualEntryType = GetCorrectTypeFlagForFormat(TarFormat.V7);
+
+            int checksum = WriteName(buffer, out _);
+            checksum += WriteCommonFields(buffer, actualLength, actualEntryType);
+            WriteChecksum(checksum, buffer);
+
+            archiveStream.Write(buffer);
+
+            if (_dataStream != null)
+            {
+                WriteData(archiveStream, _dataStream, actualLength);
+            }
+        }
+
+        // Writes the current header as a Ustar entry into the archive stream.
+        internal void WriteAsUstar(Stream archiveStream, Span<byte> buffer)
+        {
+            long actualLength = GetTotalDataBytesToWrite();
+            TarEntryType actualEntryType = GetCorrectTypeFlagForFormat(TarFormat.Ustar);
+
+            int checksum = WritePosixName(buffer);
+            checksum += WriteCommonFields(buffer, actualLength, actualEntryType);
+            checksum += WritePosixMagicAndVersion(buffer);
+            checksum += WritePosixAndGnuSharedFields(buffer);
+            WriteChecksum(checksum, buffer);
+
+            archiveStream.Write(buffer);
+
+            if (_dataStream != null)
+            {
+                WriteData(archiveStream, _dataStream, actualLength);
+            }
+        }
+
+        // Writes the current header as a PAX entry into the archive stream.
+        // Makes sure to add the preceding exteded attributes entry before the actual entry.
+        internal void WriteAsPax(Stream archiveStream, Span<byte> buffer)
+        {
+            // First, we write the preceding extended attributes header
+            TarHeader extendedAttributesHeader = default;
+            // Fill the current header's dict
+            CollectExtendedAttributesFromStandardFieldsIfNeeded();
+            // And pass them to the extended attributes header for writing
+            extendedAttributesHeader.WriteAsPaxExtendedAttributes(archiveStream, buffer, _extendedAttributes, isGea: false);
+
+            buffer.Clear(); // Reset it to reuse it
```

Comment:
These Clear calls are only necessary if the implementation assumes the data is zero'd, e.g. it doesn't write all the bytes in the buffer before reading them, assuming that anything left unwritten will be zero.  Is that the case?

Reply:
Correct, the current implementation does not write all the bytes. The tests were failing without these Clear calls, because the rented arrays had garbage from previous usages.